addons/binding/org.openhab.binding.nibeheatpump/src/main/java/org/openhab/binding/nibeheatpump/handler/NibeHeatPumpHandler.java

```diff
@@ -0,0 +1,532 @@
+        try {
+
+            if (logger.isTraceEnabled()) {
+                logger.trace("Received raw data: {}", msg.toHexString());
+            }
+
+            if (logger.isDebugEnabled()) {
+                logger.debug("Received message: {}", msg.toString());
+            }
+
+            updateStatus(ThingStatus.ONLINE);
+
+            if (msg instanceof ModbusReadResponseMessage) {
+                handleReadResponseMessage((ModbusReadResponseMessage) msg);
+            } else if (msg instanceof ModbusWriteResponseMessage) {
+                handleWriteResponseMessage((ModbusWriteResponseMessage) msg);
+            } else if (msg instanceof ModbusDataReadOutMessage) {
+                handleDataReadOutMessage((ModbusDataReadOutMessage) msg);
+            } else {
+                logger.debug("Received unknow message: {}", msg.toString());
+            }
+        } catch (Exception e) {
+            logger.debug("Error occured when parsing recevied message, reason: {}", e.getMessage());
+        }
+    }
+
+    @Override
+    public void errorOccured(String error) {
+        logger.debug("Error '{}' occured, re-establish the connection", error);
+        reconnectionRequest = true;
+        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, error);
+    }
+
+    private void handleReadResponseMessage(ModbusReadResponseMessage msg) {
+        if (readResult != null) {
+            readResult.set(msg);
+        }
+    }
+
+    private void handleWriteResponseMessage(ModbusWriteResponseMessage msg) {
+        if (writeResult != null) {
+            writeResult.set(msg);
+        }
+    }
+
+    private void handleDataReadOutMessage(ModbusDataReadOutMessage msg) {
+        List<ModbusValue> regValues = msg.getValues();
+
+        if (regValues != null) {
+            for (ModbusValue val : regValues) {
+                handleVariableUpdate(pumpModel, val);
+            }
+        }
+    }
+
+    private void handleVariableUpdate(PumpModel pumpModel, ModbusValue value) {
+        logger.debug("Received variable update: {}", value);
+        int coilAddress = value.getCoilAddress();
+
+        VariableInformation variableInfo = VariableInformation.getVariableInfo(pumpModel, coilAddress);
+
+        if (variableInfo != null) {
+            logger.trace("Using variable information to register {}: {}", coilAddress, variableInfo);
+
+            double val = (double) value.getValue() / (double) variableInfo.factor;
+            logger.debug("{} = {}", coilAddress + ":" + variableInfo.variable, val);
+
+            CacheObject oldValue = stateMap.get(coilAddress);
+            stateMap.put(coilAddress, new CacheObject(System.currentTimeMillis() / 1000, val));
+
+            if (oldValue != null && val == oldValue.value) {
+                logger.trace("Value haven't been changed, ignore update");
+            } else {
+                final String channelPrefix = (variableInfo.type == Type.Setting ? "setting#" : "sensor#");
```

Comment:
I (only) read the chapter you pointed me to, but I would like to highlight his 'conclusion'.

![image](https://user-images.githubusercontent.com/2625279/27803410-9c088220-6029-11e7-8e92-8b2687a5b37a.png)

If you read other books like "Effective Java" & "Clean Code" you see enough examples that there should not be "long or complicated methods that have many local variables". So I still think that if you code is so long and complex that you need final's to guard it, that that is the actual code smell and should be fixed instead of adding `final`'s.

But if you want it can stay 👍 

openHAB contains more of them and there are more fans 😄 see also https://github.com/eclipse/smarthome/issues/3651

PS: before Java 8 there was also a small performance benefit and it was needed to pass to anonymous inner classes, but both were improved in Java 8.